### PR TITLE
feat: add ability to use dockerfiles and images as functions

### PR DESCRIPTION
**Pull request by davemooreuws**
Relates to #447 

Things left to do (to be split into another PR):
- nitric wrapper toggle

### CODE REVIEW

File: pkg/project/options.go [hunk 1]
```diff
@@ -60,8 +63,57 @@ func FromConfig(c *Config) (*Project, error) {
 		}
 	}
 
+	for _, c := range c.Containers {
+		if c.Dockerfile != "" {
```
Comment (tjholm) [inline pkg/project/options.go:53]: Should we consider config to be invalid in this case and supply a specific error (or collection of errors).

File: pkg/project/options.go [hunk 2]
```diff
@@ -60,8 +63,57 @@ func FromConfig(c *Config) (*Project, error) {
...
+			maybeFile := filepath.Join(p.Dir, c.Dockerfile)
+
+			// Get the absolute path of docker context
+			dockerContext, err := filepath.Abs(filepath.Join(p.Dir, c.Context))
+			if err != nil {
+				return nil, err
+			}
+
+			if _, err := os.Stat(maybeFile); err != nil {
```
Comment (tjholm) [inline pkg/project/options.go]: Is a glob with no glob patterns still a valid glob in this case?

If the file actually exists can we still provide it as a glob and use the same logic branch for both actual files and glob patterns? (where an actual file would just be a single result glob pattern?)

File: pkg/provider/pulumi/common/image.go [hunk 3]
```diff
@@ -75,7 +75,9 @@ func NewImage(ctx *pulumi.Context, name string, args *ImageArgs, opts ...pulumi.
 		return nil, err
 	}
 
-	relDocker, err := filepath.Rel(args.ProjectDir, dockerFilePath)
+	dockerContext := dockerContext(args.ProjectDir, args.Compute)
+
+	relDocker, err := filepath.Rel(dockerContext, dockerFilePath)
```
Comment (tjholm) [inline pkg/provider/pulumi/common/image.go]: These pulumi builds will be deprecated in the near future, and changes related to their context probably should not be implemented (the resulting docker image doesn't/shouldn't actually copy files as it should only be wrapping an existing docker source image.

**Comment by tjholm** [comment]
LGTM, will look at merging this after:
https://github.com/nitrictech/cli/pull/453

File: pkg/project/config.go [hunk 4]
```diff
@@ -42,16 +45,86 @@ type HandlerConfig struct {
...
+	Handlers   []any          `yaml:"handlers"`
+	Containers []DockerConfig `yaml:"containers" validate:"handlers_containers_required,validate_docker_object,dive"`
+}
+
+type DockerConfig struct {
+	Dockerfile string            `yaml:"dockerfile"`
+	Image      string            `yaml:"image" validate:"omitempty,validate_docker_image"`
+	Context    string            `yaml:"context"`
+	Args       map[string]string `yaml:"args"`
+	Nitric     bool              `yaml:"nitric"`
```
Comment (tjholm) [inline pkg/project/config.go]: Can you add BaseComputeConfig here for configurable typing (will also need to be specific as `inline`.

**Comment by tjholm** [comment]
Have some new thoughts on how we can reshape this.

Instead of implementing a new concept we can probably add packaging level customisation directly to the handlers instead.

e.g.

```yaml
handlers:
  - match: './functions/*.ts'
    type: big-ol-container
    docker:
       file: ./docker/big-ol-container.dockerfile
       args:
           # Adding build args here will allow dockerfile re-use
           SOME_BUILD_ARG: some-custom-value
```

From here we will still use the same build system but we'll substitute our OOTB template with theirs, but will inject the handler references into their dockerfile the same way we do with ours, and document how to build a custom container this way.

**Comment by davemooreuws** [comment]
Converted to draft. Can be revisited later if required.

**Pull request closed without merging**